Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,485 @@
+#import "Firestore/Source/Local/FSTLocalSerializer.h"
+#import "Firestore/Source/Local/FSTLocalStore.h"
+#import "Firestore/Source/Local/FSTMemoryPersistence.h"
+#import "Firestore/Source/Local/FSTPersistence.h"
+#import "Firestore/Source/Remote/FSTSerializerBeta.h"
+#import "Firestore/Source/Util/FSTClasses.h"
+
+#include "Firestore/core/src/firebase/firestore/api/settings.h"
+#include "Firestore/core/src/firebase/firestore/auth/credentials_provider.h"
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/core/event_manager.h"
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/model/document_set.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation.h"
+#include "Firestore/core/src/firebase/firestore/remote/datastore.h"
+#include "Firestore/core/src/firebase/firestore/remote/remote_store.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/delayed_constructor.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "absl/memory/memory.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+using firestore::Error;
+using api::DocumentReference;
+using api::DocumentSnapshot;
+using api::QuerySnapshot;
+using api::Settings;
+using api::SnapshotMetadata;
+using api::ThrowIllegalState;
+using auth::CredentialsProvider;
+using auth::User;
+using core::DatabaseInfo;
+using core::ListenOptions;
+using core::EventManager;
+using core::Query;
+using core::QueryListener;
+using core::ViewSnapshot;
+using local::LruParams;
+using model::DatabaseId;
+using model::Document;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::MaybeDocument;
+using model::Mutation;
+using model::OnlineState;
+using remote::Datastore;
+using remote::RemoteStore;
+using util::Path;
+using util::AsyncQueue;
+using util::DelayedConstructor;
+using util::DelayedOperation;
+using util::Executor;
+using util::Status;
+using util::StatusCallback;
+using util::StatusOr;
+using util::StatusOrCallback;
+using util::TimerId;
+
+std::shared_ptr<FirestoreClient> FirestoreClient::Create(
+    const DatabaseInfo& database_info,
+    const api::Settings& settings,
+    std::shared_ptr<auth::CredentialsProvider> credentials_provider,
+    std::shared_ptr<util::Executor> user_executor,
+    std::shared_ptr<util::AsyncQueue> worker_queue) {
+  // Have to use `new` because `make_shared` cannot access private constructor.
+  std::shared_ptr<FirestoreClient> client(
+      new FirestoreClient(database_info, std::move(credentials_provider),
```

Comment:
You should always use `make_shared` when creating a `shared_ptr`. The way to avoid this is to use your private constructor separately, then move initialize via make_shared.

```c++
FirestoreClient client(database_info, ...);
auto client_ptr = std::make_shared<FirestoreClient>(std::move(client));
```

Reply:
The proposed change is rejected by compiler because the implicit copy constructor is deleted, due to the usage of unique_ptr<RemoteStore> as a field (If i read the error right..).

Reply:
Hmm, that implies the compiler is trying to use the copy constructor and not the move constructor, which is weird.

Reply:
This class isn't movable because it contains a `DelayedConstructor` member variable which is neither copyable nor movable.